client/client.go

```diff
@@ -159,10 +161,129 @@ func New(config *Config) (*Client, error) {
 	client.getWebsocket = func(url string) (clientWebsocket, error) {
 		return getWebsocket(transport, url)
 	}
-
+	client.Requester = &requester{
+		baseURL:   &client.baseURL,
+		userAgent: config.UserAgent,
+		client:    &http.Client{Transport: transport},
+		maintenanceHandler: func(err error) {
+			client.maintenance = err
+		},
+		warningHandler: func(count int, timestamp time.Time) {
+			client.warningCount = count
+			client.warningTimestamp = timestamp
+		},
+	}
 	return client, nil
 }
 
+type RequestOptions struct {
+	QueryString url.Values
+	Headers     map[string]string
+	Body        io.Reader
+	Output      interface{}
+}
+
+type ResponseInfo struct {
+	Type     string
+	ChangeID string
+	Result   json.RawMessage
+}
+
+type Requester interface {
+	Request(ctx context.Context, verb, path string, opts *RequestOptions) (*ResponseInfo, error)
+}
+
+type requester struct {
+	baseURL            *url.URL
+	userAgent          string
+	client             *http.Client
+	maintenanceHandler func(error)
+	warningHandler     func(count int, timestamp time.Time)
+}
+
+func (r *requester) Request(ctx context.Context, verb, path string, opts *RequestOptions) (*ResponseInfo, error) {
+	url := *r.baseURL
+	url.Path = pathpkg.Join(url.Path, path)
+	url.RawQuery = opts.QueryString.Encode()
+
+	req, err := http.NewRequestWithContext(ctx, verb, url.String(), opts.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	if r.userAgent != "" {
+		req.Header.Set("User-Agent", r.userAgent)
+	}
+	for h, v := range opts.Headers {
+		req.Header.Set(h, v)
+	}
+
+	retry := time.NewTicker(doRetry)
+	defer retry.Stop()
+	timeout := time.After(doTimeout)
```

Comment:
Should these intervals be configurable in the `RequestOptions`?

Reply:
I've not seen this value changed anywhere, but after seeing your notices PR I think we at least should provide a way to disable the retry/timeout. Will look into this